packages/lib/pod/package.json

```diff
@@ -10,6 +10,16 @@
       "types": "./dist/types/src/index.d.ts",
       "import": "./dist/esm/src/index.js",
       "require": "./dist/cjs/src/index.js"
+    },
```

Comment:
Can you talk me through the usage model of this?  I'd like to be able to reason about what we're expecting users of this library to do for their specific needs.  Do they have to choose explicitly which of these things to import, or does the build figure it out automatically?

This feels like a library-design question I should understand better before I accept.  Also something I should understand as I go about changing the serialization parts of the library.

Reply:
In this specific case, I'd like to be able to import the "podUtils" functions but without also importing `json-bigint`.

The general question is one of modularity and coupling. Importing a JavaScript module into another module (such as importing `json-bigint` into `podUtil.ts`) will, in general, ensure that any module that imports `podUtil.ts` will also import `json-bigint`. Since `index.ts` imports `podUtils.ts`, it also imports `json-bigint`.

This isn't _always_ true. In some circumstances, JavaScript bundlers can perform "dead code elimination", sometimes described as "tree-shaking", where the bundler will determine that although I'm (transitively, via `index.ts`) importing `podUtil.ts`, I never actually call any of the functions that depend on `json-bigint`. As such, those functions can be eliminated from my build, and with it the `json-bigint` dependency.

However, this isn't perfectly reliable. The `json-bigint` import might not just declare some functions or objects, it might have some side-effect that, in a non-obvious way, some other part of our code depends on. The side-effect might even be coming from some _other_ module that `json-bigint` imports! This is an annoying consequence of JavaScript allowing various kinds of global side-effect, and so bundlers will often err on the side of caution and retain code that _might_ have side-effects, just in case. Some bundlers allow manual overrides, forcing the bundler to skip the inclusion of certain dependencies when the code author knows they're not needed, but this is annoying and sometimes error-prone.

The only reliable way around this is for code to be highly modular. If there's some large-ish dependency, it should be imported in one place and the only code that is co-located with that import statement should be the code that needs to use the import. In this case, that's quite simple: only a small number of functions from `podUtil.ts` actually need `json-bigint`, so it's easy to separate them out.

But `index.ts` still depends on the new `podSerialization.ts` file, so `json-bigint` is still in the dependency graph for `index.ts`. If I do `import { checkPODName } from "@pcd/pod"`, then I'm importing `checkPODName` via `index.ts`, and this will still end up importing `podSerialization.ts` and `json-bigint`. However, we can also define "sub-path exports" for JavaScript packages, such that we can make `podUtil.ts` (in its transpiled form as `podUtil.js`) available for import. Now I can do `import { checkPODName } from "@pcd/pod/podUtil"` and the _only_ thing I am importing is the transpiled content of `podUtil.ts`.

We went through a few rounds of this when `@pcd/zuauth` was being written. `@pcd/passport-interface` has a long list of dependencies, some of which are very large in size, and so we created sub-path exports allowing pop-up window functionality to be imported by ZuAuth without importing the rest of `passport-interface`.